File: src/main/java/org/frawa/elmtest/run/ElmTestRunConfiguration.java
package org.frawa.elmtest.run;

import com.intellij.execution.Executor;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.execution.configurations.LocatableConfigurationBase;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.InvalidDataException;
import com.intellij.openapi.util.WriteExternalException;
import org.jdom.Attribute;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.nio.file.Path;
import java.nio.file.Paths;

import static org.frawa.elmtest.run.ElmTestConfigurationFactory.RUN_ICON;

public class ElmTestRunConfiguration extends LocatableConfigurationBase<ElmTestRunProfileState> {

    Options options = new Options();

    ElmTestRunConfiguration(Project project, ConfigurationFactory factory, String name) {
        super(project, factory, name);
    }

    @NotNull
    @Override
    public SettingsEditor<? extends RunConfiguration> getConfigurationEditor() {
        return new ElmTestSettingsEditor(getProject());
    }

    @Override
    public void checkConfiguration() {
    }

    @Nullable
    @Override
    public ElmTestRunProfileState getState(@NotNull Executor executor, @NotNull ExecutionEnvironment executionEnvironment) {
        return new ElmTestRunProfileState(executionEnvironment, this);
    }

    static class Options {
        String elmFolder;
    }

    @Nullable
    @Override
    public Icon getIcon() {
        return RUN_ICON;
    }


    // <ElmTestRunConfiguration elm-folder="" elm-test-binary=""/>

    static void writeOptions(Options options, Element element) {
        Element e = new Element(ElmTestRunConfiguration.class.getSimpleName());
        if (options.elmFolder != null) {
            e.setAttribute("elm-folder", options.elmFolder);
        }
        element.addContent(e);
    }

    static Options readOptions(Element element) {
        Options result = new Options();

        String name = ElmTestRunConfiguration.class.getSimpleName();
        Element optionsElement = element.getChild(name);

        if (optionsElement != null) {
            Attribute elmFolderAttr = optionsElement.getAttribute("elm-folder");
            result.elmFolder = null;
            if (elmFolderAttr != null) {
                result.elmFolder = elmFolderAttr.getValue();
            }
        }
        return result;
    }

    @Override
    public void readExternal(@NotNull Element element) throws InvalidDataException {
        this.options = readOptions(element);
    }

    @Override
    public void writeExternal(@NotNull Element element) throws WriteExternalException {
        writeOptions(this.options, element);
    }

    @Nullable
    @Override
    public String suggestedName() {
        if (options.elmFolder == null) {
            return null;
        }
        Path elmProjectName = Paths.get(options.elmFolder).getFileName();
        return elmProjectName != null ? "Tests in " + elmProjectName.toString() : null;
    }
}

File: src/main/kotlin/org/elm/ide/test/run/ElmTestRunConfiguration.kt
package org.elm.ide.test.run

import com.intellij.execution.Executor
import com.intellij.execution.configurations.ConfigurationFactory
import com.intellij.execution.configurations.LocatableConfigurationBase
import com.intellij.execution.configurations.RunConfiguration
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.InvalidDataException
import com.intellij.openapi.util.WriteExternalException
import org.elm.ide.test.run.ElmTestConfigurationFactory.Companion.RUN_ICON
import org.jdom.Element
import java.nio.file.Paths


class ElmTestRunConfiguration internal constructor(project: Project, factory: ConfigurationFactory, name: String) : LocatableConfigurationBase<ElmTestRunProfileState>(project, factory, name) {

    var options = Options()

    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration> =
            ElmTestSettingsEditor(project)

    override fun checkConfiguration() {}

    override fun getState(executor: Executor, executionEnvironment: ExecutionEnvironment) =
            ElmTestRunProfileState(executionEnvironment, this)

    class Options {
        var elmFolder: String? = null
    }

    override fun getIcon() = RUN_ICON

    @Throws(InvalidDataException::class)
    override fun readExternal(element: Element) {
        options = readOptions(element)
    }

    @Throws(WriteExternalException::class)
    override fun writeExternal(element: Element) {
        writeOptions(options, element)
    }

    override fun suggestedName(): String? {
        val elmFolder = options.elmFolder ?: return null
        return "Tests in ${Paths.get(elmFolder).fileName}"
    }

    companion object {

        // <ElmTestRunConfiguration elm-folder="" />

        fun writeOptions(options: Options, element: Element) {
            val e = Element(ElmTestRunConfiguration::class.java.simpleName)
            if (options.elmFolder != null) {
                e.setAttribute("elm-folder", options.elmFolder)
            }
            element.addContent(e)
        }

        fun readOptions(element: Element): Options {
            return Options().apply {
                val name = ElmTestRunConfiguration::class.java.simpleName
                elmFolder = element.getChild(name)?.getAttribute("elm-folder")?.value
            }
        }
    }
}
